protocols/relay/src/handler.rs

```diff
@@ -0,0 +1,550 @@
+    /// the given substream is dropped.
+    ///
+    /// Once all substreams are dropped and this handler has no other work, [`KeepAlive::Until`] can
+    /// be set eventually allowing the connection to be closed.
+    alive_lend_out_substreams: FuturesUnordered<oneshot::Receiver<()>>,
+    /// The current connection keep-alive.
+    keep_alive: KeepAlive,
+}
+
+struct OutgoingRelayReq {
+    src_peer_id: PeerId,
+    dst_peer_id: PeerId,
+    request_id: RequestId,
+    /// Addresses of the destination.
+    dst_addr: Multiaddr,
+}
+
+/// Event produced by the relay handler.
+pub enum RelayHandlerEvent {
+    /// The remote wants us to relay communications to a third party. You must either send back a
+    /// `DenyIncomingRelayReq`, or send a `OutgoingDstReq` to a different handler containing
+    /// this object.
+    IncomingRelayReq {
+        request_id: RequestId,
+        src_addr: Multiaddr,
+        req: protocol::IncomingRelayReq,
+    },
+
+    /// The remote is a relay and is relaying a connection to us. In other words, we are used as
+    /// destination. The behaviour can accept or deny the request via
+    /// [`AcceptDstReq`](RelayHandlerIn::AcceptDstReq) or
+    /// [`DenyDstReq`](RelayHandlerIn::DenyDstReq).
+    IncomingDstReq(PeerId, RequestId),
+
+    /// A `RelayReq` that has previously been sent has been accepted by the remote. Contains
+    /// a substream that communicates with the requested destination.
+    ///
+    /// > **Note**: There is no proof that we are actually communicating with the destination. An
+    /// >           encryption handshake has to be performed on top of this substream in order to
+    /// >           avoid MITM attacks.
+    OutgoingRelayReqSuccess(PeerId, RequestId, protocol::Connection),
+
+    /// The local node has accepted an incoming destination request. Contains a substream that
+    /// communicates with the source.
+    ///
+    /// > **Note**: There is no proof that we are actually communicating with the destination. An
+    /// >           encryption handshake has to be performed on top of this substream in order to
+    /// >           avoid MITM attacks.
+    IncomingDstReqSuccess {
+        stream: protocol::Connection,
+        src_peer_id: PeerId,
+        relay_addr: Multiaddr,
+    },
+
+    /// A `RelayReq` that has previously been sent by the local node has failed.
+    OutgoingRelayReqError(PeerId, RequestId),
+}
+
+/// Event that can be sent to the relay handler.
+pub enum RelayHandlerIn {
+    /// Tell the handler whether it is handling a connection used to listen for incoming relayed
+    /// connections.
+    UsedForListening(bool),
+    /// Denies a relay request sent by the node we talk to acting as a source.
+    DenyIncomingRelayReq(protocol::IncomingRelayReq),
+
+    /// Denies a destination request sent by the node we talk to.
+    DenyDstReq(PeerId, RequestId),
```

Comment:
Added a TODO note (yet to be pushed). Thank you for the hint. I am not quite sure whether there should be a separate mechanism for users to deny incoming destination requests and thus I am not sure whether this variant is needed in the first place. I can imagine that, as a relay, one would want to enforce limits on the number of incoming and outgoing relay connections. But as a destination node, the already existing connection limit mechanism should suffice. Happy for comments in case you have thoughts on this.

Reply:
To avoid confusion, I would probably remove this constructor for now, making it clear in the code that all such requests are accepted.

Reply:
With the reworked listener logic, yielding an incoming relayed connection via the correct listener (see https://github.com/libp2p/rust-libp2p/pull/1838#discussion_r582262518), this code path will no longer be unused. When receiving an incoming relayed connection but not having a listener to pass the request to, the request is denied instead.